### PR TITLE
[kwokctl] Integrate jaeger

### LINKED ISSUE #453: [kwokctl] Integrate jaeger
### What would you like to be added?

- [x] A Tracing component is built in like the built-in Prometheus component to collect tracing data from all components
- [ ] Update architecture https://kwok.sigs.k8s.io/docs/design/architecture/#kwokctl

### Why is this needed?

NONE

### PR DESCRIPTION
**Pull request by JarHMJ**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #453 

#### Does this PR introduce a user-facing change?

```release-note
[kwokctl] Integrate jaeger
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kwokctl/components/jaeger.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (wzshiming) [inline pkg/kwokctl/components/jaeger.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: pkg/kwokctl/runtime/binary/cluster.go [hunk 2]
```diff
@@ -296,6 +312,29 @@ func (c *Cluster) Install(ctx context.Context) error {
...
+
+		kubeApiserverTracingConfigData, err := k8s.BuildKubeApiserverTracingConfig(k8s.BuildKubeApiserverTracingConfigParam{
+			Endpoint: net.LocalAddress + ":" + format.String(conf.JaegerOtlpGrpcPort),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to generate kubeApiserverTracingConfig yaml: %w", err)
+		}
+		kubeApiserverTracingConfigPath = c.GetWorkdirPath(runtime.ApiserverTracingConfig)
+
+		err = os.WriteFile(kubeApiserverTracingConfigPath, []byte(kubeApiserverTracingConfigData), 0640)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/binary/cluster.go:332]:
```suggestion
		err = c.WriteFile(kubeApiserverTracingConfigPath, []byte(kubeApiserverTracingConfigData))
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 3]
```diff
@@ -317,6 +321,23 @@ func (c *Cluster) Install(ctx context.Context) error {
...
+	if conf.JaegerPort != 0 {
+		kubeApiserverTracingConfigData, err := k8s.BuildKubeApiserverTracingConfig(k8s.BuildKubeApiserverTracingConfigParam{
+			Endpoint: c.Name() + "-jaeger" + ":4317",
+		})
+		if err != nil {
+			return fmt.Errorf("failed to generate kubeApiserverTracingConfig yaml: %w", err)
+		}
+		kubeApiserverTracingConfigPath = c.GetWorkdirPath(runtime.ApiserverTracingConfig)
+
+		err = os.WriteFile(kubeApiserverTracingConfigPath, []byte(kubeApiserverTracingConfigData), 0640)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:335]:
```suggestion
		err = c.WriteFile(kubeApiserverTracingConfigPath, []byte(kubeApiserverTracingConfigData))
```

File: pkg/kwokctl/runtime/kind/cluster.go [hunk 4]
```diff
@@ -136,6 +138,22 @@ func (c *Cluster) Install(ctx context.Context) error {
...
+	if conf.JaegerPort != 0 {
+		kubeApiserverTracingConfigData, err := k8s.BuildKubeApiserverTracingConfig(k8s.BuildKubeApiserverTracingConfigParam{
+			Endpoint: conf.BindAddress + ":4317",
+		})
+		if err != nil {
+			return fmt.Errorf("failed to generate kubeApiserverTracingConfig yaml: %w", err)
+		}
+		kubeApiserverTracingConfigPath = c.GetWorkdirPath(runtime.ApiserverTracingConfig)
+
+		err = os.WriteFile(kubeApiserverTracingConfigPath, []byte(kubeApiserverTracingConfigData), 0640)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/cluster.go:151]:
```suggestion
		err = c.WriteFile(kubeApiserverTracingConfigPath, []byte(kubeApiserverTracingConfigData))
```

File: pkg/kwokctl/runtime/kind/jaeger_deployment.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/jaeger_deployment.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: pkg/utils/exec/version.go [hunk 6]
```diff
@@ -32,8 +32,14 @@ func ParseVersionFromBinary(ctx context.Context, path string) (version.Version,
 	out := bytes.NewBuffer(nil)
 	err := Exec(WithAllWriteTo(ctx, out), path, "--version")
 	if err != nil {
-		return version.Version{}, err
+		out = bytes.NewBuffer(nil)
```
Comment (wzshiming) [inline pkg/utils/exec/version.go:35]:
```suggestion
		out.Reset()
```

File: pkg/utils/exec/version.go [hunk 7]
```diff
@@ -78,7 +84,13 @@ func ParseVersionFromImage(ctx context.Context, runtime string, image string, co
 	out := bytes.NewBuffer(nil)
 	err := Exec(WithAllWriteTo(ctx, out), runtime, args...)
 	if err != nil {
-		return version.Version{}, err
+		//try match jaeger version
+		args[len(args)-1] = "version"
+		out = bytes.NewBuffer(nil)
```
Comment (wzshiming) [inline pkg/utils/exec/version.go:89]:
```suggestion
		out.Reset()
```

File: pkg/kwokctl/k8s/kube_apiserver_tracing_config.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (wzshiming) [inline pkg/kwokctl/k8s/kube_apiserver_tracing_config.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

**Review by wzshiming — CHANGES_REQUESTED** [review]
Enable jaeger on test

- https://github.com/kubernetes-sigs/kwok/blob/9929ac8f0beda4b591e1d52beba66cf9a2f063dd/test/kwokctl/kwokctl_workable_test.sh#L180
- https://github.com/kubernetes-sigs/kwok/blob/9929ac8f0beda4b591e1d52beba66cf9a2f063dd/test/kwokctl/kwokctl_with_dryrun.test.sh#L67
  - UPDATE_DRY_RUN_TESTDATA=true ./hack/e2e-test.sh  kwokctl/kwokctl_with_dryrun

File: pkg/kwokctl/cmd/create/cluster/cluster.go [hunk 9]
```diff
@@ -87,6 +88,9 @@ func NewCommand(ctx context.Context) *cobra.Command {
 `)
 	cmd.Flags().StringVar(&flags.Options.PrometheusImage, "prometheus-image", flags.Options.PrometheusImage, `Image of Prometheus, only for docker/podman/nerdctl/kind/kind-podman runtime
 '${KWOK_PROMETHEUS_IMAGE_PREFIX}/prometheus:${KWOK_PROMETHEUS_VERSION}'
+`)
+	cmd.Flags().StringVar(&flags.Options.JaegerImage, "jaeger-image", flags.Options.JaegerImage, `Image of Jaeger, only for docker/podman/nerdctl/kind/kind-podman runtime
+'${KWOK_JAEGER_IMAGE_PREFIX}/all-in-one:${KWOK_JAEGER_VERSION}'
```
Comment (wzshiming) [inline pkg/kwokctl/cmd/create/cluster/cluster.go:93]:
```suggestion
	cmd.Flags().StringVar(&flags.Options.JaegerImage, "jaeger-image", flags.Options.JaegerImage, `Image of Jaeger, only for docker/podman/nerdctl/kind/kind-podman runtime'
```

File: pkg/kwokctl/cmd/create/cluster/cluster.go [hunk 10]
```diff
@@ -87,6 +88,9 @@ func NewCommand(ctx context.Context) *cobra.Command {
...
 `)
```
Comment (wzshiming) [inline pkg/kwokctl/cmd/create/cluster/cluster.go:94]:
```suggestion
	cmd.Flags().StringVar(&flags.Options.JaegerImage, "jaeger-image", flags.Options.JaegerImage, `Image of Jaeger, only for docker/podman/nerdctl/kind/kind-podman runtime`)
```

This one doesn't implement the prefix and version from the environment variable.

File: pkg/kwokctl/components/jaeger.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		}
+	} else {
+		jaegerArgs = append(jaegerArgs,
+			"--query.http-server.host-port="+conf.BindAddress+":"+format.String(conf.Port),
+			"--collector.otlp.grpc.host-port="+net.LocalAddress+":"+format.String(conf.OtlpGrpcPort),
+		)
+	}
+
+	if conf.Verbosity != log.LevelInfo {
+		jaegerArgs = append(jaegerArgs, "--log-level="+log.ToLogSeverityLevel(conf.Verbosity))
```
Comment (wzshiming) [inline pkg/kwokctl/components/jaeger.go:66]: https://github.com/jaegertracing/jaeger/blob/2c41dae85731ee185fd31fe63f21d76fc2c8055a/pkg/es/config/config.go#L343-L351

There doesn't seem to be a `warn`, we need to define a new conversion for this

File: pkg/kwokctl/components/kube_apiserver.go [hunk 12]
```diff
@@ -80,9 +81,27 @@ func BuildKubeApiserverComponent(conf BuildKubeApiserverComponentConfig) (compon
...
+			featureGate = featureGate + "," + tracingFeatureGate
+		}
+		kubeApiserverArgs = append(kubeApiserverArgs,
+			"--feature-gates="+featureGate,
+		)
+	} else if tracingFeatureGate != "" {
 		kubeApiserverArgs = append(kubeApiserverArgs,
-			"--feature-gates="+conf.KubeFeatureGates,
+			"--feature-gates="+tracingFeatureGate,
 		)
```
Comment (wzshiming) [inline pkg/kwokctl/components/kube_apiserver.go:105]:
``` golang
	var featureGates []string
	
	if conf.KubeFeatureGates != "" {
		featureGates = append(featureGates, strings.Split(conf.KubeFeatureGates, ","))
	}
	
	...
	
	if len(featureGates) != 0 {
		kubeApiserverArgs = append(kubeApiserverArgs,
			"--feature-gates="+strings.Join(featureGates,","),
		)
	}
```

File: pkg/kwokctl/components/jaeger.go (hunk 11 above)
Comment (JarHMJ) [reply pkg/kwokctl/components/jaeger.go:66]: > https://github.com/jaegertracing/jaeger/blob/2c41dae85731ee185fd31fe63f21d76fc2c8055a/pkg/es/config/config.go#L343-L351
> 
> There doesn't seem to be a `warn`, we need to define a new conversion for this

This is `--es.log-level`.

The '--log-level' is normal.  See https://www.jaegertracing.io/docs/1.47/cli/

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 13]
```diff
@@ -317,6 +321,23 @@ func (c *Cluster) Install(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("failed to expand host volumes for kube api server component: %w", err)
 	}
+
+	kubeApiserverTracingConfigPath := ""
+	if conf.JaegerPort != 0 {
+		kubeApiserverTracingConfigData, err := k8s.BuildKubeApiserverTracingConfig(k8s.BuildKubeApiserverTracingConfigParam{
+			Endpoint: c.Name() + "-jaeger" + ":4317",
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:328]:
```suggestion
			Endpoint: c.Name() + "-jaeger:4317",
```

File: pkg/config/vars.go [hunk 14]
```diff
@@ -49,6 +49,18 @@ var (
...
+	windows = "windows"
+
+	// linux is the linux system
+	linux = "linux"
+
+	// BinarySuffixTar is the file suffix tar.gz
+	BinarySuffixTar = "tar.gz"
+
+	// BinarySuffixZip is the file suffix zip
+	BinarySuffixZip = "zip"
```
Comment (wzshiming) [inline pkg/config/vars.go:63]:
```suggestion
	// binarySuffixTar is the file suffix tar.gz
	binarySuffixTar = "tar.gz"

	// binarySuffixZip is the file suffix zip
	binarySuffixZip = "zip"
```

**Pull request merged**